nototools/generate_dingbats_html.py

```diff
@@ -385,47 +385,111 @@ def _load_fonts_and_targets(font_data, target_data, data_dir):
+    font_data.append(tuple(info))
+  return font_data
+
+
+def _read_target_data_from_file(filename):
+  """Target data uses # to indicate a comment to end of line.
+  Comments are stripped, then an empty or blank line is ignored.
+
+  Each line defines a tuple of four values: target name,
+  codelist, preferred font ids, and omitted font ids.
+  A line can also start with either of two directives,
+  !define and !default.
+
+  If a line starts with '!define ' we expect a key followed
+  by '=' and then one or more names separated by space. The
+  names are turned into a list, and entered into a dictionary
+  for the key.  Once defined a key cannot be redefined.
+
+  If a line starts with '!default ' we expect a key of either
+  'prefer' or 'omit' optionally followed by '=' and a list of
+  names to prefer or omit; these will become the default
+  values until the next '!default ' directive.  If there is
+  no '=' the value is reset.  An omitted or empty prefer or
+  omit field will get the fallback, to explicitly request None
+  and override the fallback the field should contain 'None'.
+
+  Normally, a line consists of 2-4 fields separated by ';'.
+  The first two are a target name and a codelist spec.  The
+  third is the preferred font ids separated by space,
+  previously !defined keys can be used here instead of this
+  list and the list defined for that key will be used.
+  The fourth is the omitted font ids separated by space, they
+  are treated similarly.  If the preferred or omit field is
+  missing or empty and a default value for it has been set,
+  that value is used."""
+
+  def add_index_list_or_defined(info, index, fallback, defines):
+    """Extend or update info[index], possibly using defines"""
+    if len(info) <= index:
+      info.append(fallback)
+    elif info[index] != None:
+      item = info[index]
+      if item in defines:
+        items = defines[item]
+      elif item == 'None':
+        items = None
+      elif item:
+        items = item.split()
+      else:
+        items = fallback
+      info[index] = items
+
+  prefer_fallback = None
+  omit_fallback = None
+  defines = {}
+  target_data = []
+  for line in strip_comments_from_file(filename):
+    if line.startswith('!define '):
+      name, rest = line[8:].split('=')
+      name = name.strip()
+      if name in defines:
+        raise Exception('name %s already defined in %s' % (name, filename))
+      rest = rest.strip().split()
+      defines[name] = tuple(rest)
+      continue
+    if line.startswith('!default '):
+      values = line[9:].split('=')
+      name = values[0].strip()
+      rest = values[1].strip().split() if len(values) > 1 else None
+      if not rest:
+        rest = None
+      if name == 'prefer':
+        prefer_fallback = rest
+      elif name == 'omit':
+        omit_fallback = rest
+      else:
+        raise Exception('default only understands \'prefer\' or \'omit\'')
+      continue
+    info = line.split(';')
```

Comment:
Sorry that I was not clear. Thank you for the explanation. 

I read your function-level comment and know what each field is. (but I like your explanation above  :-) ).  What I meant was  adding in-place comments would make it easier for other readers to understand.